src/basic/MAT_MAT_SHARED.hpp

```diff
@@ -0,0 +1,176 @@
+///
+///
+
+#ifndef RAJAPerf_Basic_MAT_MAT_SHARED_HPP
+#define RAJAPerf_Basic_MAT_MAT_SHARED_HPP
+
+#include "RAJA/RAJA.hpp"
+#include "common/KernelBase.hpp"
+
+#define TL_SZ 16
+
+#define MAT_MAT_SHARED_DATA_SETUP                                              \
+  Real_ptr A = m_A;                                                            \
+  Real_ptr B = m_B;                                                            \
+  Real_ptr C = m_C;
+
+/*
+ When doing the device compile pass hipcc/clang will put in the device
+ versions of the macros everywhere, in device functions, host device functions, 
+ and host only functions. Then it will make sure that code is valid everywhere, 
+ that's fine for device and host device functions, but it is not ok for host only 
+ functions. Nvcc doesn't look at host only code when it does the device pass 
+ so it doesn't see these kind of problems.
+ */
+#define MAT_MAT_SHARED_BODY_0_CLANG_HIP_CPU                   \
+  double As[TL_SZ][TL_SZ];                                    \
+  double Bs[TL_SZ][TL_SZ];                                    \
+  double Cs[TL_SZ][TL_SZ];
+
+#define MAT_MAT_SHARED_BODY_0                                                  \
+  RAJA_TEAM_SHARED double As[TL_SZ][TL_SZ];                                    \
+  RAJA_TEAM_SHARED double Bs[TL_SZ][TL_SZ];                                    \
+  RAJA_TEAM_SHARED double Cs[TL_SZ][TL_SZ];
+
+#define MAT_MAT_SHARED_BODY_1 Cs[ty][tx] = 0;
+
+#define MAT_MAT_SHARED_BODY_2                                                  \
+  const Index_type Row = by * TL_SZ + ty;                                      \
+  const Index_type Col = bx * TL_SZ + tx;                                      \
+  if (k * TL_SZ + tx < N && Row < N)                                           \
+    As[ty][tx] = A[Row * N + k * TL_SZ + tx];                                  \
+  else                                                                         \
+    As[ty][tx] = 0.0;                                                          \
+  if (k * TL_SZ + ty < N && Col < N)                                           \
+    Bs[ty][tx] = B[(k * TL_SZ + ty) * N + Col];                                \
+  else                                                                         \
+    Bs[ty][tx] = 0.0;
+
+#define MAT_MAT_SHARED_BODY_3                                                  \
+  for (Index_type n = 0; n < TL_SZ; ++n)                                       \
+    Cs[ty][tx] += As[ty][n] * Bs[n][tx];
+
+#define MAT_MAT_SHARED_BODY_4                                                  \
+  const Index_type Row = by * TL_SZ + ty;                                      \
+  const Index_type Col = bx * TL_SZ + tx;                                      \
+  if (Row < N && Col < N)                                                      \
+    C[Col + N * Row] = Cs[ty][tx];
+
+using launch_policy = RAJA::expt::LaunchPolicy<RAJA::expt::seq_launch_t
```

Comment:
Something to think about: is it better/clearer to put the policy definitions in the header file or in the source files for the various back-end variants?  All the other kernels have them in the source files. I see why you put them here. But, it may be better to put them in the source files. That would avoid the need for macro guards in the launch policies, which can make them harder to read.

Reply:
@rhornung67 , I moved some of the policies to the source files; however for the CPU based back-ends I still needed to add macro guards; this is on account of  the possibility of compiling with or without GPU support. Something @MrBurmark and I have chatted about is offering a way to specify a single policy even when compiling with the GPU -- this could clean up adding kernels to the perfsuite.